synapse/storage/state.py

```diff
@@ -48,6 +50,179 @@ def __len__(self):
         return len(self.delta_ids) if self.delta_ids else 0
 
 
+@attr.s(slots=True)
+class StateFilter(object):
+    """A filter used when querying for state.
+
+    The default, `StateFilter()`, is equivalent to no filter.
+
+    Attributes:
+        types (Iterable[str, str|None]|None): list of 2-tuples of the form
+            (`type`, `state_key`), where a `state_key` of `None` matches all
+            state_keys for the `type`. If None, all types are returned.
+        filtered_types(list[str]|None): Only apply filtering via `types` to this
+            list of event types.  Other types of events are returned unfiltered.
+            If None, `types` filtering is applied to all events.
+    """
+
+    types = attr.ib(default=None)
+    filtered_types = attr.ib(default=None)
+
+    def get_member_split(self):
+        """Split the filter into member vs non-member types.
+
+        Returns:
+            tuple[Iterable[str, str|None]|None]: Returns a tuple of member,
+            non-member types in the same format as `types`.
+        """
+        if self.types is not None:
+            non_member_types = [t for t in self.types if t[0] != EventTypes.Member]
+
+            if (self.filtered_types is not None
+                    and EventTypes.Member not in self.filtered_types):
+                # we want all of the membership events
+                member_types = None
+            else:
+                # Check if there is a ("m.room.member", None) entry, if there
+                # is we can just return None.
+                if any(s is None for t, s in self.types if t == EventTypes.Member):
+                    member_types = None
+                else:
+                    member_types = [t for t in self.types if t[0] == EventTypes.Member]
+
+        else:
+            non_member_types = None
+            member_types = None
+
+        return member_types, non_member_types
+
+    def return_expanded(self):
```

Comment:
I think the comment [below](https://github.com/matrix-org/synapse/pull/4011/files#diff-6ee0c965f89a99e084e8500580e3ec17R908) about why this is a useful thing to do now belongs up here somewhere.